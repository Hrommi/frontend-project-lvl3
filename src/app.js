import * as yup from 'yup';
import axios from 'axios';
import uniqueId from 'lodash/uniqueId';
import i18next from 'i18next';
import watch from './watchers';
import parse from './parser';

const CORS_PROXY_URL = 'https://cors-anywhere.herokuapp.com';
const FETCHING_TIMEOUT = 5000;

const urlSchema = yup
  .string()
  .url(() => i18next.t('form.errors.url'))
  .required(() => i18next.t('form.errors.required'));

const validateUrl = (url, feeds) => {
  try {
    urlSchema.validateSync(url);
    if (feeds.find((feed) => feed.url === url)) {
      throw new Error(i18next.t('form.errors.exist'));
    }
    return null;
  } catch (e) {
    return e.message;
  }
};

const handleUrlInput = (e, state) => {
  const {
    target: { value },
  } = e;
  const { form, feeds } = state;
  form.url = value;
  form.feedback.value = null;

  if (value === '') {
    form.valid = true;
    return;
  }

  const urlError = validateUrl(value, feeds);
  if (urlError) {
    form.valid = false;
  } else {
    form.valid = true;
  }
};

const fetchPosts = (state) => {
  const { feeds, posts } = state;
  const promises = feeds.map((feed) => (
    axios.get(`${CORS_PROXY_URL}/${feed.url}`).then((response) => {
      const parsedFeed = parse(response.data);
      const newPosts = parsedFeed.items.map((item) => ({
        ...item,
        feedId: feed.id,
      }));
      const oldPostsLinks = new Set(
        posts.filter((post) => post.feedId === feed.id).map((post) => post.link),
      );
      const differencePosts = newPosts.filter(
        (newPost) => !oldPostsLinks.has(newPost.link),
      );
      posts.unshift(...differencePosts);
    })
  ));

  Promise.all(promises).finally(() => {
    setTimeout(() => fetchPosts(state), FETCHING_TIMEOUT);
  });
};

const handleFormSubmit = (e, state) => {
  e.preventDefault();
  const { form, feeds, posts } = state;
  form.state = 'loading';
  const urlError = validateUrl(form.url, feeds);
  if (!urlError) {
    form.feedback.value = null;
    axios
      .get(`${CORS_PROXY_URL}/${form.url}`)
      .then((response) => {
        const feed = parse(response.data);
        const feedId = uniqueId();
        posts.push(
          ...feed.items.map((item) => ({
            ...item,
            feedId,
          })),
        );
        feeds.unshift({
          id: feedId,
          title: feed.title,
          description: feed.description,
          url: form.url,
        });
        form.url = '';
        form.state = 'filling';
        form.feedback.value = i18next.t('form.success');
        form.feedback.type = 'success';
      })
      .catch((error) => {
        form.state = 'failed';
        form.feedback.value = error.message;
        form.feedback.type = 'danger';
      });
  } else {
    form.state = 'failed';
    form.feedback.value = urlError;
    form.feedback.type = 'danger';
  }
};

export default () => {
  const state = {
    form: {
      state: 'filling',
      url: '',
      valid: true,
      feedback: {
        value: null,
        type: 'danger',
      },
    },
    feeds: [],
    posts: [],
  };

  const elements = {
    form: document.querySelector('.rss-form'),
    urlInput: document.querySelector('.rss-form [name="url"]'),
    submitButton: document.querySelector('.rss-form [type="submit"]'),
    spinner: document.querySelector('.rss-form .spinner'),
    feedback: document.querySelector('.feedback'),
    feeds: document.querySelector('.feeds'),
  };

  const watchedState = watch(state, elements);

  elements.urlInput.addEventListener('input', (e) => {
    handleUrlInput(e, watchedState);
  });

  elements.form.addEventListener('submit', (e) => {
    handleFormSubmit(e, watchedState);
  });

  setTimeout(() => fetchPosts(watchedState), FETCHING_TIMEOUT);
};
