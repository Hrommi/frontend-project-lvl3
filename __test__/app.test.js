import { promises as fs } from 'fs';
import path from 'path';
import i18next from 'i18next';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import app, { CORS_PROXY_URL } from '../src/app';

const routes = {
  feed: () => 'http://localhost.com/feed.rss',
  updateFeed: () => 'http://localhost.com/updated-feed.rss',
  notFound: () => 'http://localhost.com/404',
};

const getFixturePath = (name) => path.resolve(__dirname, '__fixtures__', name);

let formElement;
let urlInputElement;
let feedbackElement;
let feedsElement;

beforeEach(async () => {
  const pathToHtml = getFixturePath('index.html');
  const html = await fs.readFile(pathToHtml, 'utf8');
  document.body.innerHTML = html;

  formElement = document.querySelector('.rss-form');
  urlInputElement = document.querySelector('.rss-form [name="url"]');
  feedbackElement = document.querySelector('.feedback');
  feedsElement = document.querySelector('.feeds');

  const data = await fs.readFile(getFixturePath('correctRss.xml'), 'utf-8');
  const updatedData = await fs.readFile(getFixturePath('updatedCorrectRss.xml'), 'utf-8');
  const mock = new MockAdapter(axios);
  mock
    .onGet(`${CORS_PROXY_URL}/${routes.feed()}`)
    .reply(200, data)
    .onGet(`${CORS_PROXY_URL}/${routes.updateFeed()}`)
    .reply(200, updatedData)
    .onGet(`${CORS_PROXY_URL}/${routes.notFound()}`)
    .reply(404);

  app();
});

describe('app', () => {
  test('should initialize correctly', () => {
    expect(true).toBeDefined();
  });

  test('should show an error message for an empty url', async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    formElement.dispatchEvent(new Event('submit'));
    expect(feedbackElement.innerHTML).toBe(i18next.t('form.errors.required'));
  });

  test('should show an error message for an invalid url', async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    urlInputElement.value = 'test';
    urlInputElement.dispatchEvent(new Event('input'));
    formElement.dispatchEvent(new Event('submit'));
    expect(feedbackElement.innerHTML).toBe(i18next.t('form.errors.url'));
  });

  test('should show an error message with the 404 url', async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    urlInputElement.value = routes.notFound();
    urlInputElement.dispatchEvent(new Event('input'));
    formElement.dispatchEvent(new Event('submit'));
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(feedbackElement.innerHTML).toBe('Request failed with status code 404');
  });

  test('should render feed', async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    urlInputElement.value = routes.feed();
    urlInputElement.dispatchEvent(new Event('input'));
    formElement.dispatchEvent(new Event('submit'));
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(feedbackElement.innerHTML).toBe(i18next.t('form.success'));
    expect(feedsElement).toMatchSnapshot();
  });

  test('should render two feeds', async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    urlInputElement.value = routes.feed();
    urlInputElement.dispatchEvent(new Event('input'));
    formElement.dispatchEvent(new Event('submit'));
    await new Promise((resolve) => setTimeout(resolve, 0));
    urlInputElement.value = routes.updateFeed();
    urlInputElement.dispatchEvent(new Event('input'));
    formElement.dispatchEvent(new Event('submit'));
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(feedbackElement.innerHTML).toBe(i18next.t('form.success'));
    expect(feedsElement).toMatchSnapshot();
  });
});
