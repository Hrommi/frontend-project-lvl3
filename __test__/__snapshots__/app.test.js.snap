// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should render feed 1`] = `
<div
  class="col-md-10 col-lg-8 mx-auto feeds"
>
  <div
    class="mb-4"
  >
    <h2>
      <a
        href="http://localhost.com/feed.rss"
        target="_blank"
      >
        RSS Title
      </a>
    </h2>
    <p>
      RSS Description
    </p>
    <div>
      <a
        href="https://rss.com/1"
      >
        Title 1
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/2"
      >
        Title 2
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/3"
      >
        Title 3
      </a>
    </div>
  </div>
</div>
`;

exports[`app should render two feeds 1`] = `
<div
  class="col-md-10 col-lg-8 mx-auto feeds"
>
  <div
    class="mb-4"
  >
    <h2>
      <a
        href="http://localhost.com/updated-feed.rss"
        target="_blank"
      >
        RSS Title
      </a>
    </h2>
    <p>
      RSS Description
    </p>
    <div>
      <a
        href="https://rss.com/0"
      >
        Title 0
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/1"
      >
        Title 1
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/2"
      >
        Title 2
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/3"
      >
        Title 3
      </a>
    </div>
  </div>
  

  <div
    class="mb-4"
  >
    <h2>
      <a
        href="http://localhost.com/feed.rss"
        target="_blank"
      >
        RSS Title
      </a>
    </h2>
    <p>
      RSS Description
    </p>
    <div>
      <a
        href="https://rss.com/1"
      >
        Title 1
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/2"
      >
        Title 2
      </a>
    </div>
    

    <div>
      <a
        href="https://rss.com/3"
      >
        Title 3
      </a>
    </div>
  </div>
</div>
`;
